feat(push-notifications): Open link in current tab if possible

sorin-davidoi opened the pull request:
Does not break functionality if the website is added to the home screen.

cc: @beatrix-bitrot

beatrix-bitrot commented:
not working for me on desktop chrome but i will test now on mobile...

sorin-davidoi commented:
@beatrix-bitrot Refresh,  close all Mastodon tabs and then open a new one. Service workers do not take control of the clients immediately.

sorin-davidoi commented:
Is working for me (on glitch.social).

beatrix-bitrot commented:
Working on mobile. Won't be able to test desktop for a while

beatrix-bitrot commented:
Actually an issue I have now... It doesn't open the thing the notification is about... Just... Some arbitrary one from before.... But it's the same one every time?!

sorin-davidoi commented:
Cannot reproduce :cry:

Gargron commented on app/javascript/mastodon/service_worker/web_push_notifications.js:
```diff
@@ -75,7 +87,7 @@ const handleNotificationClick = (event) => {
     } else {
       event.notification.close();
-      resolve(self.clients.openWindow(event.notification.data.url));
+      resolve(openUrl(event.notification.data.url));
```
I don't know how handleNotificationClick is attached exactly, but I remember reading that React events are synthetic and are reused between calls. So perhaps the thing @beatrix-bitrot is experiencing is event.notification.data.url referring to something from an earlier call. You could try making a copy of the data you're interested in outside of the Promise.

sorin-davidoi commented:
This is code that runs in the Service Worker, where you don't have access to the DOM - no React, no DOM APIs. Will try to reproduce this again later today.

sorin-davidoi commented:
Still not able to reproduce. I've changed a bit the logic that determines which client is focused (e.g. if one tab is on `/web` and another on `/settings`, open the link in `/web`). I've also added some code to skip waiting when a new service worker is installed - this is perfectly fine since we do not have to deal with caches and assets - although Nolan will probably change this in his PR.

sorin-davidoi commented on app/javascript/mastodon/service_worker/web_push_notifications.js:
```diff
@@ -50,6 +50,24 @@ const makeRequest = (notification, action) =>
+    if (clientList.length !== 0 && 'navigate' in clientList[0]) { // Chrome 42-48 does not support navigate
+      const webClients = clientList
+        .filter(client => /\/web\//.test(client.url))
+        .sort(client => client !== 'visible');
```
If there are multiple webClients, if one is visible it should be the first.

sorin-davidoi commented on app/javascript/mastodon/service_worker/entry.js:
```diff
@@ -1 +1,10 @@
+
+// Cause a new version of a registered Service Worker to replace an existing one
+// that is already installed, and replace the currently active worker on open pages.
+self.addEventListener('install', function(event) {
```
@nolan Am I right in assuming that this is safe to do since we do not cache and there is no communication between the clients and the Service Worker?

nolanlawson replied:
Yes `skipWaiting` should be fine in this case. Since we're not caching, there's no situation I can think of where the "frontend" would get out of sync with the "backend" (service worker).
